### PR TITLE
Harden error paths and add HTTP timeouts on external connections

### PR DESCRIPTION
**Pull request by DIvyaNautiyal07**
## Summary

Addresses the P0 items in #59. Five resilience fixes across AgentDB, the Fireworks LLM fallback, the Olas subgraph, and the tweepy / twikit connections.

> **Update (review-round 1):** Reworked AgentDB to raise on non-2xx/non-404, fixed a silent no-op in the tweepy adapter, dropped `wait_on_rate_limit`, added per-call timeout overrides, and tightened tests. See per-comment replies for what changed and why.

### What's fixed

**1. AgentDB JSON guard + write-failure contract** — `packages/valory/skills/agent_db_abci/`
`_request` previously parsed the body with an unwrapped `json.loads` and raised `RuntimeError` on any status that wasn't 200/201/404. Both paths propagated to the framework tick loop, which only catches `StopIteration`, so the agent crash-restart-looped. Now: malformed JSON on a 2xx returns `None` with an error log; non-success / non-404 status codes raise `RuntimeError` so write paths can distinguish "row missing" from "server broken" and avoid duplicate-write hazards. The boundary is `agents_fun_db.py`, which catches `RuntimeError` for write paths (`add_interaction`, `update_twitter_details`) and the read aggregator (`get_tweet_feedback`). `LoadDatabaseBehaviour.async_act` catches and yields the period so the FSM does not crash-loop.

**2. Fireworks JSON guard** — `packages/valory/skills/memeooorr_abci/behaviour_classes/base.py:1007`
The alternative-LLM fallback parsed `response.body` with an unwrapped `json.loads`. Now wrapped, returns `None` on `JSONDecodeError`.

**3. Olas subgraph schema guard** — same file, around line 665
`get_memeooorr_handles_from_subgraph` did `services["units"]`, which would `KeyError` on schema drift. Now `services.get("units")`, returns `[]` on missing/None, and logs a warning so a schema drift is distinguishable from "no services registered".

**4. tweepy request timeout** — `packages/valory/connections/tweepy/`
`tweepy.Client` previously held a `requests.Session` with no timeout. `tweepy.API` already passed `timeout=60` so the indefinite-hang risk was concentrated on `tweepy.Client`. With `MAX_WORKER_THREADS=1`, a single hung Client request blocked every subsequent tweepy call. Mounts a `_TimeoutHTTPAdapter` on both sessions for both `http://` and `https://`, default 30s, configurable via `tweepy_request_timeout`. The adapter checks `kwargs.get("timeout") is None` rather than `setdefault` because `Session.request()` always populates the kwarg — `setdefault` was a silent no-op. `wait_on_rate_limit` is intentionally **not** enabled: tweepy implements it as `time.sleep(reset_at - now)` (up to ~15min) which the per-request adapter cannot bound and would freeze the connection on a single 429.

**Coverage gap (acknowledged):** the adapter mount only patches `Client.session` and `API.session`. tweepy's `OAuth*Handler` token-fetch and the rate-limit handler use their own `requests` paths and are not patched here. These are startup-only, not the steady-state hung-call concern this PR targets.

**5. twikit request timeout** — `packages/valory/connections/twikit/`
Same `MAX_WORKER_THREADS=1` problem. Adds a `_with_timeout(coro, name, timeout=None)` helper that wraps twikit calls in `asyncio.wait_for`. Default 30s for short-lived ops; **per-call override** to 120s for `upload_media` (Twitter video transcoding routinely takes 30-90s with `wait_for_completion=True`) and the two `client.login` calls (cold-start FlowToken handshake + email/captcha challenge). Bare call sites (`search`, `get_user_tweets`, `get_user_by_screen_name`) catch `asyncio.TimeoutError` and return safe empty values (`[]`, `[]`, `{}`) so a timeout does not propagate uncaught into the dialogue handler.

### Tests

- AgentDB — `test_non_success_status_raises_runtime_error`, `test_malformed_json_body_returns_none`. New `agents_fun_db.py` boundary tests for `RuntimeError` swallowing on `add_interaction`, `update_twitter_details`, and `get_tweet_feedback`.
- Subgraph — `test_returns_empty_list_when_units_key_is_absent` (uses a non-empty dict so the new branch is actually exercised, not the early-empty short-circuit), `test_returns_empty_list_when_units_value_is_none`. Both assert the warning is emitted.
- Fireworks — `test_returns_none_when_response_body_is_not_json`.
- tweepy — `TestTwitterTimeoutHardening`: adapter mounted on both sessions for both schemes; **`test_adapter_injects_timeout_when_session_request_omits_it`** drives the real `Session.request()` path so a regression to `setdefault` would fail; `Client` constructed without `wait_on_rate_limit`.
- twikit — `TestWithTimeout`: returns within budget, raises `asyncio.TimeoutError` when exceeded, logs the configured number of seconds (`f"{timeout}s"`); `search()` / `get_user_tweets()` / `get_user_by_screen_name()` return safe empty values on timeout; per-call `timeout=` override propagates to the log line.

### Out of scope

P1, P2, P3, P4 items in #59 will land in follow-up PRs. P3 #10 (circuit breaker) will be filed as a separate design issue.

## Test plan
- [ ] `tox -e black-check,isort-check,flake8,mypy,pylint,darglint`
- [ ] `tox -e check-packages,check-hash,check-abciapp-specs,check-dependencies`
- [ ] `autonomy packages lock --check`
- [ ] `tox -e py3.12-linux`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/valory/skills/agent_db_abci/agent_db_client.py [hunk 1]
```diff
@@ -189,12 +189,22 @@ def _request(
...
         if response.status_code in [200, 201]:
-            return json.loads(response.body)
+            try:
+                return json.loads(response.body)
+            except (json.JSONDecodeError, ValueError):
+                self.logger.error(
+                    f"AgentDB returned non-JSON body for {method} {url}: "
+                    f"{response.body!r}"
+                )
+                return None
```
Comment (OjusWiZard) [inline packages/valory/skills/agent_db_abci/agent_db_client.py:199]: **Critical — silent corruption risk (confidence 92).**

A 200 OK with malformed body now resolves to `None`, the *same* sentinel callers already use to mean "not found". A successful-but-corrupt response is a server contract violation, not "doesn't exist" — collapsing them is dangerous for the get-then-create pattern (e.g. `get_agent_type_by_type_name` → `create_agent_type` at L223–215): a corrupt body on GET makes the caller think the row is missing and POST a duplicate.

Suggested: raise on JSON parse failure of a 2xx, or return a typed sentinel distinct from 404.

File: packages/valory/skills/agent_db_abci/agent_db_client.py [hunk 2]
```diff
@@ -189,12 +189,22 @@ def _request(
...
 
         if response.status_code == 404:
             return None
 
-        raise RuntimeError(f"Request failed: {response.status_code} - {response.text}")
+        self.logger.warning(
+            f"AgentDB request failed: {response.status_code} - {response.text}"
+        )
+        return None
```
Comment (OjusWiZard) [inline packages/valory/skills/agent_db_abci/agent_db_client.py]: **Critical — duplicate-write risk (confidence 92).**

This collapses 4xx/5xx, auth failures, and network errors into the same `None` sentinel as 404. Caller hazards across this file:

- `get_agent_type_by_type_name` / `get_agent_instance_by_address`: callers follow get-then-create. A transient 500 on GET → `create_agent_*` runs → **duplicate registry rows.**
- `create_agent_type` (L215), `create_agent_instance` (L250), `create_attribute_instance`, `update_attribute_instance`: if the server *committed* but returned 5xx mid-response, caller sees `None`, retries next round → **duplicate writes.**
- `delete_*` paths: 5xx silently looks like success-with-empty-body — the row may not have been deleted.

`logger.warning` is not a substitute for surfacing a write failure to the FSM. **Recommend:** keep raising for non-404 non-2xx (or return a typed Result/sentinel) so write paths can distinguish "not found" from "server broken" and decide whether to retry.

File: packages/valory/connections/tweepy/tweepy_wrapper.py [hunk 3]
```diff
@@ -69,9 +99,13 @@ def __init__(  # pylint: disable=too-many-arguments, too-many-positional-argumen
             consumer_secret=consumer_secret,
             access_token=access_token,
             access_token_secret=access_token_secret,
+            wait_on_rate_limit=True,
```
Comment (OjusWiZard) [inline packages/valory/connections/tweepy/tweepy_wrapper.py]: **Important — liveness regression (confidence 85).**

`wait_on_rate_limit=True` makes tweepy `time.sleep` up to ~15 minutes on 429s. The `_TimeoutHTTPAdapter` mounted below only bounds a single HTTP request — it does **not** bound this inter-request sleep. Net effect: a single `post_tweet` can block the connection for a full rate-limit window where it previously raised, which is bad for an ABCI agent with round-time SLAs.

Options: (a) keep `wait_on_rate_limit=False` and handle 429 explicitly with bounded backoff; (b) wrap calls in an outer asyncio/threading timeout above the tweepy layer.

File: packages/valory/connections/tweepy/tweepy_wrapper.py [hunk 4]
```diff
@@ -69,9 +99,13 @@ def __init__(  # pylint: disable=too-many-arguments, too-many-positional-argumen
...
         )
         self.logger = logger if logger else DEFAULT_LOGGER
 
+        _mount_timeout_adapter(getattr(self.client, "session", None), request_timeout)
+        _mount_timeout_adapter(getattr(self.api, "session", None), request_timeout)
```
Comment (OjusWiZard) [inline packages/valory/connections/tweepy/tweepy_wrapper.py:123]: **Important — adapter coverage gap (confidence 82).**

`getattr(..., "session", None)` only patches `client.session` and `api.session`. tweepy's `OAuth*Handler` token-fetch and the internal rate-limit-handler use their own `requests` paths and remain unbounded.

The commit message claims a 30s default timeout on every request — that holds for steady-state calls, not auth/token negotiation. Either document the gap or extend the patch (e.g., monkeypatch `requests.adapters.HTTPAdapter.send` defaults at module init).

**Review by OjusWiZard — COMMENTED** [review]
Automated multi-perspective review (code-reviewer + silent-failure-hunter). Confidence ≥ 80 only. Two **critical** findings that, IMO, invert the PR's intent (loud failure → silent corruption) — flagged inline. Tests in this PR codify the new lossy contracts; if you accept the critical findings, those tests need updating.

File: packages/valory/connections/twikit/connection.py [hunk 5]
```diff
@@ -339,14 +342,27 @@ async def _get_response(  # pylint: disable=too-many-return-statements
...
 
+    async def _with_timeout(self, coro: Any, name: str) -> Any:
+        """Run a twikit coroutine with the configured request timeout."""
+        try:
+            return await asyncio.wait_for(coro, timeout=self.request_timeout)
+        except asyncio.TimeoutError:
+            self.logger.error(
+                f"twikit {name} did not respond within {self.request_timeout}s"
+            )
+            raise
```
Comment (OjusWiZard) [inline packages/valory/connections/twikit/connection.py:368]: **Important — `TimeoutError` escapes existing twikit retry loops (confidence 90).**

Re-raising `asyncio.TimeoutError` here means it propagates past existing retry loops in this file, because they only catch `twikit.errors.TwitterException` (which `TimeoutError` is **not** a subclass of):

- `create_tweet` (L488), `delete_tweet` (L527), `verify_tweet` (L508), `like_tweet`, `follow_user`, `retweet`: first timeout terminates the dialogue handler instead of triggering `MAX_POST_RETRIES` / `MAX_GET_RETRIES`.
- Bare call sites with no surrounding try (`search_tweets` L433, `get_user_tweets` L541–545, `get_user_by_screen_name` L617): timeout propagates uncaught to `_handle_send`. Previously these hung; now they crash.

This worsens the failure mode the PR is supposed to bound. Fix: either broaden the retry loops to also catch `asyncio.TimeoutError`, or have `_with_timeout` log + return a sentinel and let each caller decide.

File: packages/valory/skills/memeooorr_abci/behaviour_classes/base.py [hunk 6]
```diff
@@ -662,7 +662,7 @@ def get_memeooorr_handles_from_subgraph(self) -> Generator[None, None, List[str]
         if not services:
             return handles
 
-        for service in services["units"]:
+        for service in services.get("units") or []:
```
Comment (OjusWiZard) [inline packages/valory/skills/memeooorr_abci/behaviour_classes/base.py]: **Important — silent schema-drift (confidence 80).**

`services.get("units") or []` correctly avoids `KeyError`, but emits no log. If the subgraph schema actually drifts (the stated motivation), the agent silently produces an empty handle list and proceeds as if no services exist — a silent functional outage rather than a guarded one.

Suggestion: log a `warning` when `services` is truthy but `units` is missing/None — distinguishes "no services" from "schema broken".

File: packages/valory/skills/memeooorr_abci/behaviour_classes/base.py [hunk 7]
```diff
@@ -1004,7 +1004,13 @@ def replace_tweet_with_alternative_model(
...
 
         # Load the response
-        api_data = json.loads(response.body)
+        try:
+            api_data = json.loads(response.body)
+        except (json.JSONDecodeError, ValueError):
+            self.context.logger.error(
+                f"Fireworks returned non-JSON body: {response.body!r}"
+            )
+            return None
```
Comment (OjusWiZard) [inline packages/valory/skills/memeooorr_abci/behaviour_classes/base.py:1025]: **Medium — overloaded `None` sentinel (confidence 80).**

`replace_tweet_with_alternative_model` already uses `None` for legitimate "alternative model declined / unavailable" elsewhere. With this guard, a Fireworks outage is now indistinguishable from a decline.

Non-blocking, but worth a distinct sentinel if upstream logic should retry on outage but accept on decline.

**Review by LOCKhart07 — COMMENTED** [review]
## Findings

### Blocking

**B1. `_TimeoutHTTPAdapter.send` is a silent no-op for `tweepy.Client` — the central fix doesn't actually fire.**
`packages/valory/connections/tweepy/tweepy_wrapper.py:49`

```python
kwargs.setdefault("timeout", self._timeout)
```

`requests.Session.request()` always populates `kwargs["timeout"]` before reaching the adapter — with `None` when the caller omits it. `tweepy.Client.request()` does **not** pass a timeout (verified against the installed tweepy: `Client.request` calls `self.session.request(method, host+route, params=..., json=..., headers=..., auth=...)` with no `timeout`). So `kwargs["timeout"] is None`, `setdefault` is a no-op, and the adapter's 30s default is never applied. Reproduced with the exact adapter from this PR mounted on a real `Session` — `HTTPAdapter.send` receives `timeout=None`.

(`tweepy.API.request` happens to pass `timeout=self.timeout`, defaulting to 60s, so the API session was never the infinite-hang risk and the adapter has no effect there either.)

Fix:

```python
if kwargs.get("timeout") is None:
    kwargs["timeout"] = self._timeout
```

**B2. The unit test that should have caught B1 masks it.**
`packages/valory/connections/tweepy/tests/test_tweepy_wrapper.py:246` (`test_adapter_injects_timeout_when_caller_omits_it`)

The test calls `adapter.send(MagicMock())` directly with empty kwargs. Production traffic always goes through `Session.send`, which puts `timeout=None` into kwargs — that's the path the adapter has to override. Drive the real path: construct a `Session`, mount the adapter, patch `HTTPAdapter.send` to capture kwargs, then call `s.request("GET", url)` and assert the captured `timeout == 30`. Today this test passes; with B1 fixed it should still pass; without B1 fixed but with the right test, it should fail.

**B3. Subgraph "missing units key" test doesn't exercise the new code path.**
`packages/valory/skills/memeooorr_abci/tests/test_behaviour_classes_base.py:1050` (`test_returns_empty_list_when_units_key_is_absent`)

The fake returns `{}`. But `get_memeooorr_handles_from_subgraph` already has an `if not services: return handles` short-circuit that catches the empty dict before reaching the new `services.get("units") or []`. The test would pass without the new fix — it's verifying the pre-existing guard. Use a non-empty dict missing the key instead, e.g. `return {"foo": "bar"}`. The sibling `test_returns_empty_list_when_units_value_is_none` (returns `{"units": None}`) is correct.

### Non-blocking

**N1. `wait_on_rate_limit=True` blocks the single tweepy worker for up to ~15 min on 429.**
`packages/valory/connections/tweepy/tweepy_wrapper.py:102`

tweepy implements this as `time.sleep(reset_at - now)` between requests. With `MAX_WORKER_THREADS=1`, every subsequent tweepy call on the connection is blocked until the sleep completes, and the timeout adapter can't shorten it (the sleep is outside the HTTP call). Probably acceptable vs. failing every method on a 429, but the PR description doesn't surface this tradeoff.

**N2. PR description nit.**
The description says the tweepy `Client` and `API` sessions "had no timeout (the requests default is `None` = infinite)." `tweepy.API.timeout` defaults to 60s and is passed explicitly to `session.request`, so `tweepy.API` was already finite — the infinite-hang risk was only on `tweepy.Client`. Worth tightening for the linked #59 paper trail.

**Review by DhairyaPatel7 — DISMISSED** [review]

File: packages/valory/connections/twikit/connection.py [hunk 8]
```diff
@@ -616,8 +659,11 @@ async def upload_media(self, media_path: str) -> Optional[str]:
                     )
 
                 # Upload media to Twitter
-                result = await self.client.upload_media(
-                    source=media_path, wait_for_completion=True
+                result = await self._with_timeout(
+                    self.client.upload_media(
+                        source=media_path, wait_for_completion=True
+                    ),
+                    "upload_media",
```
Comment (DhairyaPatel7) [inline packages/valory/connections/twikit/connection.py:695]: **Important — uniform 30s timeout is too tight for long-running ops (confidence 80).**

`twikit_request_timeout` is a single budget applied to all 13 wrapped calls. Two of them have legitimately long runtimes:

- `upload_media(..., wait_for_completion=True)` (here): for videos, Twitter's transcoding routinely takes 30–90s; with a 30s budget this will consistently `TimeoutError` on video posts.
- `client.login(...)` (L389, L410): the FlowToken handshake plus any email/captcha challenge can exceed 30s on first login.

Combined with the earlier point that `asyncio.TimeoutError` escapes existing twikit retry loops (it isn't a `TwitterException`), video uploads and cold-start logins are likely to fail outright at the default. Suggest either raising the default well above 30s (e.g. 120s) or letting `_with_timeout` accept a per-call override and using a longer budget for `upload_media` and `login` specifically.

File: packages/valory/connections/twikit/tests/test_connection.py [hunk 9]
```diff
@@ -1375,4 +1377,57 @@ async def test_delete_tweet_all_retries_exhausted(self) -> None:
...
+
+        async def slow() -> None:
+            await asyncio.sleep(60)
+
+        with pytest.raises(asyncio.TimeoutError):
+            await conn._with_timeout(slow(), "search_tweet")
+
+        message = conn.logger.error.call_args.args[0]
+        assert "search_tweet" in message
+        assert "0" in message  # the configured timeout is included
```
Comment (DhairyaPatel7) [inline packages/valory/connections/twikit/tests/test_connection.py]: **Nit — assertion is too loose.**

`"0" in message` matches any substring `"0"` (an error code, an ID, a path). If the format string changes to e.g. `"timed out (code 1023)"` this still passes. Tighten to something like `assert "0s" in message` or `assert f"{conn.request_timeout}s" in message`.

File: packages/valory/skills/agent_db_abci/agent_db_client.py (hunk 1 above)
Comment (DIvyaNautiyal07) [reply packages/valory/skills/agent_db_abci/agent_db_client.py:199]: Confirmed and fixed in 1351e15f.

Switched to the contract you suggested: 2xx with malformed body returns None and logs an error, but non-2xx / non-404 now raises RuntimeError again. Get-then-create paths can now distinguish a real 404 from a transient outage.

The catch boundary moved up one level into agents_fun_db.py (and LoadDatabaseBehaviour.async_act for the load path), so callers don't have to be updated individually but the FSM still doesn't crash-loop.

File: packages/valory/skills/agent_db_abci/agent_db_client.py (hunk 2 above)
Comment (DIvyaNautiyal07) [reply packages/valory/skills/agent_db_abci/agent_db_client.py]: Confirmed and fixed in 1351e15f.

Restored the RuntimeError on non-200/201/404 so write paths see a clear signal. Caught at the agents_fun_db.py boundary:

- `add_interaction`: catches RuntimeError and returns None (matches the existing failure contract — callers already check for None and skip post-write logic).
- `update_twitter_details`: catches and logs; the next period will retry the update.
- `get_tweet_feedback`: returns `{"likes": 0, "retweets": 0, "replies": []}` on RuntimeError so the read path degrades instead of crashing.

For the load path, `LoadDatabaseBehaviour.async_act` catches RuntimeError and returns from the period without sending a payload; the round timeout fires and the next period retries.

Net effect: 5xx no longer creates the "row missing → create duplicate" hazard you flagged, and the FSM doesn't crash-loop either.

File: packages/valory/connections/tweepy/tweepy_wrapper.py (hunk 3 above)
Comment (DIvyaNautiyal07) [reply packages/valory/connections/tweepy/tweepy_wrapper.py]: Confirmed and reverted in 1351e15f.

Dropped `wait_on_rate_limit=True` on `tweepy.Client`. With `MAX_WORKER_THREADS=1` the inter-request `time.sleep` would freeze the connection for up to a full rate-limit window, which is worse than the original "fail immediately and let the FSM move on" shape. 429 surfaces as `tweepy.errors.TooManyRequests` again; bounded retry on 429 belongs in the connection layer, not blanket sleep.

Updated the test from asserting `wait_on_rate_limit is True` to `is False/None` and added a docstring on the `Client` construction explaining the choice.

File: packages/valory/connections/tweepy/tweepy_wrapper.py (hunk 4 above)
Comment (DIvyaNautiyal07) [reply packages/valory/connections/tweepy/tweepy_wrapper.py:123]: Acknowledged, deferring to a follow-up.

The mount only patches `Client.session` and `API.session`. `OAuth*Handler` token-fetch and the rate-limit handler use their own `requests` paths and remain unbounded.

Reasoning for not landing this in the same PR: the auth handlers fire once at startup (token negotiation, OAuth dance), not in the steady-state. The audit's #4 concern ("a single hung response indefinitely blocks ALL subsequent tweepy operations") is the steady-state read/write path on the connection's worker thread, which the current adapter does cover. Patching `requests.adapters.HTTPAdapter.send` defaults at module init is invasive and worth its own PR rather than slipping into the resilience patch.

I've added an explicit "Coverage gap (acknowledged)" section to the PR description so this isn't lost. Happy to file a follow-up issue if you'd like.

File: packages/valory/connections/twikit/connection.py (hunk 5 above)
Comment (DIvyaNautiyal07) [reply packages/valory/connections/twikit/connection.py:368]: Partial agree, fixed the real issue in 1351e15f.

Verified the existing retry loops for `create_tweet` (post_tweet body), `delete_tweet`, `upload_media`, and `validate_login`: they catch `except Exception`, so `asyncio.TimeoutError` does fall into the retry. But you're right about the bare call sites — `search`, `get_user_tweets`, and `get_user_by_screen_name` had no surrounding try/except, so a timeout would propagate uncaught into the dialogue handler.

Wrapped each bare call site explicitly:
- `search`: catches asyncio.TimeoutError and returns `[]`.
- `get_user_tweets`: returns `[]` if either the user-lookup or the tweet-fetch times out.
- `get_user_by_screen_name`: returns `{}` on timeout.

Tests cover all three (`test_search_returns_empty_list_on_timeout`, `test_get_user_tweets_returns_empty_list_on_timeout`, `test_get_user_by_screen_name_returns_empty_dict_on_timeout`). The previously-hanging path now resolves to a safe empty value instead of either hanging or crashing.

File: packages/valory/skills/memeooorr_abci/behaviour_classes/base.py (hunk 6 above)
Comment (DIvyaNautiyal07) [reply packages/valory/skills/memeooorr_abci/behaviour_classes/base.py]: Confirmed and fixed in 1351e15f.

Added an explicit warning when `services` is truthy but `units` is missing or None. The branch now logs `"Olas subgraph returned no 'units' in payload; treating as empty handle list. Payload keys: [...]"` so an actual schema drift surfaces in ops dashboards distinctly from the legitimate "no services registered" path (the early `if not services` short-circuit, which doesn't log).

Both schema-drift tests now assert `logger.warning.assert_called_once()`. I also fixed a separate gap: the existing "missing units" test was using `{}` which short-circuits at the early empty guard before reaching the new code. Switched to a non-empty dict so the new branch is actually exercised.

File: packages/valory/skills/memeooorr_abci/behaviour_classes/base.py (hunk 7 above)
Comment (DIvyaNautiyal07) [reply packages/valory/skills/memeooorr_abci/behaviour_classes/base.py:1025]: Acknowledged, leaving as-is.

Fair point that `None` here collapses "Fireworks declined" with "Fireworks unreachable". I considered a distinct sentinel but the upstream handler treats either case the same way — fall back to the original tweet, no retry on this path — so the distinction wouldn't change behaviour today.

Filing a typed sentinel here would mean a contract change with no consumer that benefits, and I'd rather not add it just because it's possible. If/when an upstream wants to discriminate (e.g. retry on outage but accept on decline), that's the right time to introduce a 3-state result.

File: packages/valory/connections/twikit/connection.py (hunk 8 above)
Comment (DIvyaNautiyal07) [reply packages/valory/connections/twikit/connection.py:695]: Confirmed and fixed in 1351e15f.

Added a per-call `timeout` override on `_with_timeout` and a class-level `LONG_OPERATION_TIMEOUT_SECONDS = 120` for the calls you flagged:

- `upload_media(..., wait_for_completion=True)` — 120s, covers Twitter video transcoding
- both `client.login(...)` call sites in `twikit_login` — 120s, covers FlowToken handshake + email/captcha challenge on cold start

Default for everything else stays at 30s (configurable via `twikit_request_timeout`). The `asyncio.TimeoutError` escape concern from the sibling thread is also resolved: `upload_media`'s existing `while retries < MAX_POST_RETRIES: try ... except Exception` catches `TimeoutError` so a flaky transcode now retries cleanly within the 120s budget per attempt.

Test: `test_with_timeout_accepts_per_call_override` asserts the override flows through to the wait_for budget and the log line.

File: packages/valory/connections/twikit/tests/test_connection.py (hunk 9 above)
Comment (DIvyaNautiyal07) [reply packages/valory/connections/twikit/tests/test_connection.py]: Confirmed and fixed in 1351e15f.

Tightened the assertion to `f"{conn.request_timeout}s" in message`, and bumped the test's configured timeout from 0 to 7 so a stray "0" in some other part of the log line can't accidentally match. The format string is `"twikit {name} did not respond within {budget}s"`, so "7s" must appear verbatim — a future change to the log format that drops the seconds suffix would now fail the test.

**Comment by DIvyaNautiyal07** [comment]
@LOCKhart07 thanks for the catch on B1 — that was the most important finding in the review. Status on each:

**B1 — adapter `setdefault` no-op.** Fixed in 1351e15f. Replaced `kwargs.setdefault('timeout', self._timeout)` with an explicit `if kwargs.get('timeout') is None` check. Verified your reproduction shape: `Session.request()` always populates `kwargs['timeout']=None` before reaching the adapter, so `setdefault` was indeed a silent no-op for any caller (including `tweepy.Client.request`) that didn't pass a timeout explicitly. The 30s default now actually fires.

**B2 — test masks B1.** Fixed in the same commit. Replaced `test_adapter_injects_timeout_when_caller_omits_it` with `test_adapter_injects_timeout_when_session_request_omits_it`, which mounts the adapter on a real `requests.Session`, patches `HTTPAdapter.send`, and calls `session.request('GET', url)` so the captured kwargs go through the production code path. A regression to `setdefault` would now fail.

**B3 — subgraph test short-circuits.** Fixed. The fake now returns `{'foo': 'bar'}` instead of `{}`, so the early `if not services: return handles` guard is bypassed and the new `services.get('units') or []` branch is actually exercised. Sibling `test_returns_empty_list_when_units_value_is_none` was already correct. Both tests now also assert that the schema-drift warning is emitted.

**N1 — wait_on_rate_limit liveness.** Reverted to default (`False`) in 1351e15f. Reasoning matches yours: with `MAX_WORKER_THREADS=1` the inter-request `time.sleep` would freeze the entire connection on a 429, which the per-request adapter cannot bound. Surfacing 429 as `tweepy.errors.TooManyRequests` is the original behaviour and is the right shape; bounded retry on 429 belongs in the connection layer, not blanket sleep. The new behaviour is documented in a comment on the `Client(...)` construction.

**N2 — PR description nit.** Corrected. The description now says only `tweepy.Client` was the infinite-hang risk; `tweepy.API.timeout` already defaulted to 60s so it was finite. Also added an explicit "Coverage gap (acknowledged)" section noting that auth handlers / OAuth token-fetch paths are not yet patched (deferred to a follow-up since they're startup-only).

Full local CI is green on the new commit.

File: packages/valory/skills/memeooorr_abci/behaviour_classes/db.py [hunk 10]
```diff
@@ -44,16 +44,28 @@ def async_act(self) -> Generator:
...
+                # AgentDB returned a non-success status. Rather than
+                # crashing the FSM (which would supervisor-restart into
+                # the same crash loop), log and yield without sending a
+                # payload. The round timeout will fire and the next
+                # period will retry the load.
+                self.context.logger.error(
+                    f"AgentDB unavailable during LoadDatabaseRound: {exc}. "
+                    "Skipping period to retry on the next round."
+                )
+                return
```
Comment (LOCKhart07) [inline packages/valory/skills/memeooorr_abci/behaviour_classes/db.py:66]: **Non-blocking — early `return` tight-loops on AgentDB outage (confidence 80).**

This path returns without `set_done()` and without `yield from self.sleep(...)`. The framework's `act()` re-creates the generator on the next tick, which immediately re-runs `load_db()` → another `_request` → another `RuntimeError`. With AgentDB returning 5xx in millisecond range, that's many retries per second until the round timeout fires (~60s).

The "round timeout will fire and the next period will retry" reasoning is correct for FSM recovery, but it doesn't bound the request rate during the wait — only the eventual transition.

Suggest:
```python
except RuntimeError as exc:
    self.context.logger.error(...)
    yield from self.sleep(self.params.sleep_time)
    return
```
The `yield from self.sleep(...)` pattern is already used in `chain.py:549` and `twitter.py:1309`.

File: packages/valory/skills/memeooorr_abci/behaviour_classes/db.py [hunk 11]
```diff
@@ -44,16 +44,28 @@ def async_act(self) -> Generator:
...
+                    persona,
+                    hearting_cooldown_hours,
+                    summon_cooldown_seconds,
+                ) = yield from self.load_db()
+                yield from self.populate_keys_in_kv()
+                yield from self.init_own_twitter_details()
+                agent_details = self.gather_agent_details(persona)
+
+                yield from self._write_kv({"agent_details": agent_details})
+            except RuntimeError as exc:
```
Comment (LOCKhart07) [inline packages/valory/skills/memeooorr_abci/behaviour_classes/db.py:58]: **Nit — coverage gap: no test for this branch.**

`agents_fun_db.py` got `test_add_interaction_returns_none_on_runtime_error`, `test_update_twitter_details_swallows_runtime_error`, and `test_feedback_returns_safe_default_on_runtime_error` for the boundary catches, but the FSM-layer catch added here has no corresponding test in `test_behaviour_classes_db.py::TestAsyncAct`. Worth a test that asserts `set_done()` is **not** called and no payload is sent when `load_db` raises `RuntimeError`, so a future refactor that drops this guard fails loudly.

File: packages/valory/skills/memeooorr_abci/behaviour_classes/db.py (hunk 10 above)
Comment (DIvyaNautiyal07) [reply packages/valory/skills/memeooorr_abci/behaviour_classes/db.py:66]: Confirmed and fixed in b32c9647.

Verified your reading of the framework — `act()` re-creates the generator on each tick, so a bare `return` after the catch lets a 5xx-ing AgentDB get hit many times per second until the round timeout. Now:

```python
except RuntimeError as exc:
    self.context.logger.error(
        f"AgentDB unavailable during LoadDatabaseRound: {exc}"
    )
    yield from self.sleep(self.params.sleep_time)
    return
```

`self.params.sleep_time` is provided by `abstract_round_abci` via `models.py:266` and configured to `1` in `memeooorr_abci/skill.yaml:168`, matching the pattern in `chain.py:549`.

Test `test_async_act_skips_period_when_load_db_raises_runtime_error` asserts that `self.sleep` is called exactly once with `sleep_time` on the catch path.

File: packages/valory/skills/memeooorr_abci/behaviour_classes/db.py (hunk 11 above)
Comment (DIvyaNautiyal07) [reply packages/valory/skills/memeooorr_abci/behaviour_classes/db.py:58]: Confirmed and fixed in b32c9647.

New `test_async_act_skips_period_when_load_db_raises_runtime_error` in `TestAsyncAct`:

```python
behaviour.load_db = mock_load_db  # raises RuntimeError("AgentDB request failed: 500")
behaviour.sleep = mock_sleep      # records calls
behaviour.send_a2a_transaction = mock_send_a2a_transaction
behaviour.set_done = MagicMock()

# drive the generator...

assert not payloads_sent
behaviour.set_done.assert_not_called()
assert sleep_calls == [behaviour.params.sleep_time]
behaviour.context.logger.error.assert_called_once()
```

Asserts that the catch path (a) sends no payload, (b) does not call `set_done`, (c) sleeps exactly once with the configured `sleep_time`, and (d) logs the error. A future refactor that drops the guard would fail at least one of these. Coverage report now reports 100% on `db.py` (was 94% with lines 58-68 missing).

**Review by LOCKhart07 — APPROVED** [review]

**Review by OjusWiZard — APPROVED** [review]
